sources/packages/backend/apps/workers/src/controllers/disbursement/_tests_/e2e/disbursement.controller.associateMSFAA.e2e-spec.ts

```diff
@@ -0,0 +1,556 @@
+      },
+    );
+
+    // Create student and save fake application with disbursements.
+    const application = await saveFakeApplicationDisbursements(
+      db.dataSource,
+      { student },
+      {
+        offeringIntensity: OfferingIntensity.partTime,
+        applicationStatus: ApplicationStatus.Completed,
+        createSecondDisbursement: true,
+      },
+    );
+
+    // Act
+    const associateMSFAAPayload = createFakeAssociateMSFAAPayload({
+      assessmentId: application.currentAssessment.id,
+    });
+    const saveResult = await disbursementController.associateMSFAA(
+      associateMSFAAPayload,
+    );
+
+    // Asserts
+    expect(FakeWorkerJobResult.getResultType(saveResult)).toBe(
+      MockedZeebeJobResult.Complete,
+    );
+
+    // Fetch MSFAA Number for the student in SIMS.
+    const assignedMSFAANumber = await db.msfaaNumber.findOne({
+      select: {
+        id: true,
+        msfaaNumber: true,
+        dateSigned: true,
+        dateRequested: true,
+        serviceProviderReceivedDate: true,
+        referenceApplication: { id: true },
+      },
+      relations: {
+        referenceApplication: true,
+      },
+      where: {
+        student: { id: student.id },
+      },
+    });
+    // Assert MSFAA Number.
+    expect(assignedMSFAANumber.msfaaNumber).toBe(savedMSFAANumber.msfaaNumber);
+    // Assert date signed.
+    expect(assignedMSFAANumber.dateSigned).toBe(savedMSFAANumber.dateSigned);
+    // Assert date requested.
+    expect(assignedMSFAANumber.dateRequested).not.toBe(null);
+    // Assert service provider date.
+    expect(assignedMSFAANumber.serviceProviderReceivedDate).not.toBe(null);
+  });
+
+  it("Should create new MSFAA Number for part time application when MSFAA is found for previously signed disbursement and its offering end date is not between the valid date.", async () => {
+    // Arrange
+    const student = await saveFakeStudent(db.dataSource);
+    // Create and save fake MSFAA Number.
+    const savedMSFAANumber = await db.msfaaNumber.save(
+      createFakeMSFAANumber(
+        {
+          student: student,
+        },
+        {
+          msfaaState: MSFAAStates.Signed,
+          msfaaInitialValues: {
+            offeringIntensity: OfferingIntensity.partTime,
+            dateSigned: getISODateOnlyString(addDays(-731)),
+          },
+        },
+      ),
+    );
+    // Create previous fake application with disbursements.
+    await saveFakeApplicationDisbursements(
+      db.dataSource,
+      { student, msfaaNumber: savedMSFAANumber },
+      {
+        offeringIntensity: OfferingIntensity.partTime,
+        applicationStatus: ApplicationStatus.Completed,
+        firstDisbursementInitialValues: {
+          disbursementScheduleStatus: DisbursementScheduleStatus.Sent,
+        },
+        offeringInitialValues: {
+          studyStartDate: getISODateOnlyString(addDays(-1000)),
+          studyEndDate: getISODateOnlyString(addDays(-731)),
```

Comment:
It can be replaced by `MAX_MSFAA_VALID_DAYS + 1` and `1000` can be replaced by `MAX_MSFAA_VALID_DAYS + 2`.